## Задание

На сайте запущен А/В тест с целью увеличить доход. В приложенном excel файле вы найдете сырые данные по результатам 
эксперимента – user_id, тип выборки variant_name и доход принесенный пользователем revenue.
Проанализируйте результаты эксперимента и напишите свои рекомендации менеджеру.


In [51]:
from typing import Union
from tqdm import tqdm

import pandas as pd
import numpy as np

from scipy import stats
from statsmodels.stats.meta_analysis import effectsize_smd
from statsmodels.stats import proportion
from statsmodels.stats.power import tt_ind_solve_power
from statsmodels.stats.power import zt_ind_solve_power

In [64]:
df = pd.read_excel('AB_Test_Results.xlsx')

df.head(10)

C:\anaconda3\lib\site-packages\openpyxl\worksheet\_read_only.py:79: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


USER_ID VARIANT_NAME  REVENUE
0      737      variant      0.0
1     2423      control      0.0
2     9411      control      0.0
3     7311      control      0.0
4     6174      variant      0.0
5     2380      variant      0.0
6     2849      control      0.0
7     9168      control      0.0
8     6205      variant      0.0
9     7548      control      0.0

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   USER_ID       10000 non-null  int64  
 1   VARIANT_NAME  10000 non-null  object 
 2   REVENUE       10000 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 234.5+ KB


In [66]:
df.shape

(10000, 3)

In [67]:
df.USER_ID.nunique()

6324

In [68]:
df = df.groupby(['USER_ID', 'VARIANT_NAME'], as_index=False).agg({'REVENUE': 'sum'})
df

USER_ID VARIANT_NAME  REVENUE
0           2      control     0.00
1           3      control     0.00
2           3      variant     0.00
3           4      variant     0.00
4           5      variant     0.00
...       ...          ...      ...
7860     9996      control     0.00
7861     9996      variant     6.46
7862     9998      control     0.00
7863    10000      control     0.00
7864    10000      variant     0.00

[7865 rows x 3 columns]

In [69]:
df.shape

(7865, 3)

In [70]:
df.groupby('USER_ID', as_index=False).agg({'VARIANT_NAME': 'count'})['VARIANT_NAME'].value_counts()

1    4783
2    1541
Name: VARIANT_NAME, dtype: int64

In [71]:
unique_ids = \
(df
 .groupby('USER_ID', as_index=False)
 .agg({'VARIANT_NAME': 'count'})
 #.['VARIANT_NAME'].value_counts()
 .query('VARIANT_NAME == 1')
 .USER_ID
 .values
 )

In [72]:
df_new = df[df.USER_ID.isin(unique_ids)].copy(deep=True)
df

USER_ID VARIANT_NAME  REVENUE
0           2      control     0.00
1           3      control     0.00
2           3      variant     0.00
3           4      variant     0.00
4           5      variant     0.00
...       ...          ...      ...
7860     9996      control     0.00
7861     9996      variant     6.46
7862     9998      control     0.00
7863    10000      control     0.00
7864    10000      variant     0.00

[7865 rows x 3 columns]

In [73]:
df_new.describe()

USER_ID      REVENUE
count  4783.000000  4783.000000
mean   4994.395777     0.135873
std    2898.618472     3.011392
min       2.000000     0.000000
25%    2476.000000     0.000000
50%    4975.000000     0.000000
75%    7515.000000     0.000000
max    9998.000000   196.010000

In [83]:
control = df_new[df_new['VARIANT_NAME'] == 'control'].copy(deep=True)
treatment = df_new[df_new['VARIANT_NAME'] == 'variant'].copy(deep=True)

In [84]:
control.describe()

USER_ID      REVENUE
count  2390.000000  2390.000000
mean   5020.881590     0.196887
std    2904.850992     4.172201
min       2.000000     0.000000
25%    2517.250000     0.000000
50%    5012.500000     0.000000
75%    7616.000000     0.000000
max    9998.000000   196.010000

In [85]:
treatment.describe()

USER_ID      REVENUE
count  2393.000000  2393.000000
mean   4967.943168     0.074935
std    2892.745368     0.858207
min       4.000000     0.000000
25%    2435.000000     0.000000
50%    4955.000000     0.000000
75%    7379.000000     0.000000
max    9995.000000    23.040000

In [86]:
def continious_result(control: pd.DataFrame,
                      treatment: pd.DataFrame,
                      column: str,
                      n_iters: int = 10_000) -> pd.DataFrame:
    # Статистика по выборкам
    size = control.loc[:, column].shape[0]
    
    control_mean = control.loc[:, column].mean()
    treatment_mean = treatment.loc[:, column].mean()
    
    control_std = control.loc[:, column].std(ddof=1)
    treatment_std = treatment.loc[:, column].std(ddof=1)
    
    # Бутсрап
    booted_diff = []
    for _ in tqdm(range(n_iters)):
        control_sample = control.loc[:, column].sample(n=size, replace=True).values
        treatment_sample = treatment.loc[:, column].sample(n=size, replace=True).values
        booted_diff.append(np.mean(control_sample - treatment_sample))
    
    # Считаем статистику после бустрапа
    md_ci, std_ci = np.mean(booted_diff), np.std(booted_diff, ddof=1)
    left_ci, right_ci = np.percentile(booted_diff, [2.5, 97.5])
    p_value_ci = 2 * (1 - stats.norm.cdf(np.abs(md_ci / std_ci)))
    
    # Считаем мощность эксперимента
    effect_size, _ = effectsize_smd(mean1=treatment_mean, sd1=treatment_std, nobs1=size,
                                    mean2=control_mean, sd2=control_std, nobs2=size)
    power = tt_ind_solve_power(effect_size=effect_size,
                               nobs1=size,
                               alpha=.05,
                               power=None,
                               ratio=1)
    # Формируем отчёт 
    result = pd.DataFrame({'effect_size': effect_size,
                           'alpha': p_value_ci, 
                           'beta': (1-power),
                           'CI': f'[{np.round(left_ci, 3)}, {np.round(right_ci, 3)}]',
                           'difference': md_ci,},
                          index=[column]) 
    return result

In [93]:
import plotly.express as px
fig = px.histogram(df,
                   x='REVENUE',
                   color = 'VARIANT_NAME',
                   title='REVENUE',
                   marginal = 'box',
                   nbins = 100,
                   barmode='overlay')

fig.show()

In [94]:
continious_result(control, treatment, column='REVENUE')

100%|██████████████████████████████████████████████████████████████████████████| 10000/10000 [00:03<00:00, 3135.69it/s]


effect_size     alpha      beta               CI  difference
REVENUE    -0.040483  0.160954  0.712143  [-0.003, 0.324]    0.122487

Ответ: доверительный  интервал [-0.003, 0.324] включает ноль, значение alpha приевышает 0,05 (0.160954), beta приевышает 0,20 (0.712143), результат стастистически не значим. Рекомендации менеджеру: либо надо переделывать тест, либо продолжить, отложив на другое время, чтобы отследить эффект.	